core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java

```diff
@@ -302,22 +309,23 @@ private boolean tryRehash()
         int[] newGroupIds = new int[newCapacity];
         Arrays.fill(newGroupIds, -1);
 
-        for (int groupId = 0; groupId < nextGroupId; groupId++) {
-            if (groupId == nullGroupId) {
-                continue;
-            }
-            long value = valuesByGroupId.get(groupId);
+        for (int i = 0; i < values.length; i++) {
```

Comment:
Why do we do it? We iterate by values instead of "groups". Then,we have one more sequential random access in the 314 line?

Reply:
I mean random. I cannot understand why we change it.

Reply:
Previous approach was iterating over a group and then jumping around both `groupIds` and `values` arrays which was completely random.
Now we iterate over those two arrays in sequential order. Also `newGroupIds` whill be traversed quasi-sequentailly since it follow the same order as `groupIds` - it is ordered by hash.
Reach out to me offline if you need a more robust description.

Reply:
Yeah, I get it.